### DISCUSSION TITLE
BaseSettings - username as member variable changes unintended

### DISCUSSION THREAD

**Discussion by tbobek** [pydantic/pydantic#4488] [Question] [answered]
During the development of a python-based MQTT service I use pydantic type checking and models. I like it very much as I normally develop in Go or C++ and there one always know what types are used. 
However, I experienced a strange behaviour today on a linux machine, where my username is `thomas`. I condensed it to a smallest possible example to demonstrate the issue. Firstly, I import pydantic's `BaseSettings` class and then derive a class called `Test` from it, filling it with some values, among them `username`. When I print the contents of the class, I see that the username has changed to my login name. Explicit setting of this member variable *after* initialization changes it finally (second line of output). 

My question would be: **is this the intended behaviour or a bug?** I don't find it very intuitive. It took me a while to identify why I couln't connect to the MQTT broker with a configuration having a username different from my login name ... 

I'd appreciate to discuss this, maybe there is a good reason for this behaviour that I don't see at the moment. 

*Source:*
```
from pydantic import BaseSettings

class Test(BaseSettings): 
    name: str = "testname"
    value: int = 42
    username: str = "bla"


t = Test()


print("Test = {}".format(t.dict()))
t.username = "blubb"
print("Test = {}".format(t.dict()))
```

*Output* after calling `python test.py`:
```
Test = {'name': 'testname', 'value': 42, 'username': 'thomas'}
Test = {'name': 'testname', 'value': 42, 'username': 'blubb'}
```

**Comment by samuelcolvin** [comment, marked as answer]
`BaseSettings` child classes reads environments variables, see [the docs](https://pydantic-docs.helpmanual.io/usage/settings/).

Your OS will have set the environment variable `username=thomas`, hence the field being filled.

If you don't want to read environment variables, don't use `BaseSettings`.

In V2, the environment variable source will be optional on `BaseSettings` child classes, but it's not optional in V1.X.

**Reply by tbobek** [reply]
Thanks for the quick answer. if I insert an `os.getenv("username", "nn")` into the script I get "nn" back, so I guess the environment variable `username` is not set. 

Perhaps, the variable `$USER` is read, which is set in Linux to the name of the current user. 

I appreciate that it will be optional in V2. Thanks again!

**Reply by samuelcolvin** [reply]
`BaseSettings` definitely reads env variables based on the field name or `env`, as per the docs. :shrug:

**Reply by samuelcolvin** [reply]
oh, it's case insensitive by default, so it might be be that the env var is `USERNAME`.
